src/platform/cert_capi_openssl.c

```diff
@@ -0,0 +1,266 @@
+
+#include "platform_internal.h"
+
+#define OPENSSL_SUPPRESS_DEPRECATED 1 // For hmac.h, which was deprecated in 3.0
+#include "openssl/err.h"
+#include "openssl/hmac.h"
+#include "openssl/kdf.h"
+#include "openssl/pem.h"
+#include "openssl/rsa.h"
+#include "openssl/ssl.h"
+#include "openssl/x509.h"
+#ifdef QUIC_CLOG
+#include "cert_capi_openssl.c.clog.h"
+#endif
+
+#ifdef _WIN32
+#include <wincrypt.h>
+#include <msquic.h>
+
+QUIC_STATUS
+QuicTlsExtractPrivateKey(
+    _In_ const QUIC_CREDENTIAL_CONFIG* CredConfig,
+    _Out_ RSA** RsaKey,
+    _Out_ X509** X509Cert
+    )
+{
+    QUIC_CERTIFICATE* Cert = NULL;
+    BYTE* KeyData = NULL;
+    RSA* Rsa = NULL;
+    DWORD KeyLength = 0;
+    NCRYPT_KEY_HANDLE KeyHandle = 0;
+    PCCERT_CONTEXT CertCtx = NULL;
+    X509* X509CertStorage = NULL;
+    DWORD ExportPolicyProperty = 0;
+    DWORD ExportPolicyLength = 0;
+    QUIC_STATUS Status;
+    int Ret = 0;
+
+    if (QUIC_FAILED(
+        Status =
+            QuicCertCreate(CredConfig, &Cert))) {
+        QuicTraceEvent(
+            LibraryErrorStatus,
+            "[ lib] ERROR, %u, %s.",
+            Status,
+            "QuicCertCreate");
+        goto Exit;
+    }
+
+    CertCtx = (PCCERT_CONTEXT)Cert;
+    X509CertStorage =
+        d2i_X509(
+            NULL,
+            (const unsigned char**)&CertCtx->pbCertEncoded,
+            CertCtx->cbCertEncoded);
+    if (X509CertStorage == NULL) {
+        QuicTraceEvent(
+            LibraryError,
+            "[ lib] ERROR, %s.",
+            "d2i_X509 failed");
+        Status = QUIC_STATUS_OUT_OF_MEMORY;
+        goto Exit;
+    }
+
+    KeyHandle = (NCRYPT_KEY_HANDLE)QuicCertGetPrivateKey(Cert);
+    if (KeyHandle == 0) {
+        Status = QUIC_STATUS_INTERNAL_ERROR;
+        goto Exit;
+    }
+
+    if (FAILED(Status =
+        NCryptGetProperty(
+            KeyHandle,
+            NCRYPT_EXPORT_POLICY_PROPERTY,
+            (PBYTE)&ExportPolicyProperty,
+            sizeof(ExportPolicyProperty),
+            &ExportPolicyLength, 0))) {
+        QuicTraceEvent(
+            LibraryError,
+            "[ lib] ERROR, %s.",
+            "NCryptGetProperty failed");
+        goto Exit;
+    }
+
+    if ((ExportPolicyProperty & NCRYPT_ALLOW_PLAINTEXT_EXPORT_FLAG) == 0) {
+        QuicTraceEvent(
+            LibraryError,
+            "[ lib] ERROR, %s.",
+            "Requested certificate does not support exporting. An exportable certificate is required");
+        Status = QUIC_STATUS_INVALID_PARAMETER;
+        goto Exit;
+    }
+
+    if (FAILED(
+        Status =
+            NCryptExportKey(
+                KeyHandle,
+                0,
+                BCRYPT_RSAFULLPRIVATE_BLOB,
+                NULL,
+                NULL,
+                0,
+                &KeyLength,
+                0))) {
+        QuicTraceEvent(
+            LibraryErrorStatus,
+            "[ lib] ERROR, %u, %s.",
+            Status,
+            "NCryptExportKey failed. Need exportable certificate");
```

Comment:
Is "Need exportable certificate" really needed? Is that the _only_ reason this can fail? Especially considering you have the check for export flag above?

Reply:
Ditto below.